keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+  * Minimize review burden on other SIGs
+  * Leave logging in acceptable state if effort fails
+* Simplify ingestion of logs into third party logging solutions by adding an option to output logs in the JSON format
+
+
+### Non-Goals
+
+* We are **not** replacing currently used logging library (klog) or the way in which it is used
+* We are **not** preserving full compatibility of logging output
+* We are **not** proposing structuring all logs in Kubernetes
+
+## Proposal
+
+This KEP proposes introducing a standardized log message structure. To enforce a new log message structure we are proposing adding new methods to the klog library: `InfoS`, `WarningS` , `ErrorS`  and `FatalS` which will provide a structured interface for building log messages and creating new helper methods which will provide consistent identification of Kubernetes objects in logs.
+
+With the new methods we would like also to migrate to klog v2 and introduce a new JSON logging output format to Kubernetes components that would be an alternative to the current text format and which will make querying and processing Kubernetes logs even simpler.
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
+
+We would like for the Kubernetes community to settle on one preferred log message structure. Proposed structure should:
+* Separate log message from its arguments
+* Treat log arguments as key-value pairs
+* Be easily parsable and queryable
+* Have a specific guidance on log message and its arguments
+
+For this purpose we suggest to use following log message structure:
+```
+<message> <key1>=<value1> <key2>=<value2> ...
+```
+
+where
+* message is formatted using “%q” fmt logic
+* keys are formatted using “%s” fmt logic
+* values are formatted using “%v” fmt logic and optionally escaped using the “%q” fmt logic if it not a number or a boolean
+* keys are separated from values with a “=” sign
+* message and key-values are separated with a single space
+
+#### New klog methods
+
+To enforce a new log message structure we want to introduce new methods to klog library which will provide a more structured interface for formatting log messages compared to current methods based on fmt format strings.
+For each format method (`Infof`, `Warningf`, `Errorf`, `Fatalf`) we will add matching structured method (`InfoS`, `WarningS`, `ErrorS`, `FatalS`).
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] “Pod status updated“ pod=“kubedns“ status=“ready“
```

Comment:
eventually I assume we would want the line header to also be reflected as key=value items in the log line, I think it's worth mentioning this as a consideration once everything has been migrated

Reply:
As there is already a lot of logging agents which have a regex for parsing this header I believe that changing this would be quite disruptive and would break a lot of clients. Therefore if there is an interest in changing this I would recommend implementing this as a separate output format.

Reply:
yeah I think this would be of interest but only as a later step, so I don't think we necessarily have to decide now